### ISSUE TITLE
[QUESTION] Reusing Athena's Parquet reader

### ISSUE THREAD

**Issue by danielhaviv** [awslabs/aws-athena-query-federation#129]
Hi,
I'm implementing a metadata handler that returns a list of parquet files as the splits.
Is it possible to read Parquet files using Athena's own reader or do I have to implement my own reader?

TY.
Daniel

**Comment by avirtuos** [comment]
This is a feature that is on our radar, essentially to allow you to implement only the MetadataHandler and to use Athena's internal S3 read capability to any of the supported formats. Unfortunately, for now you'll have to do the parquet reads yourself. The good news is that it is very easy to go from parquet to apache arrow using arrow's built in load from parquet feature.

I'll leave this issue open and update it to a feature request.

**Comment by danielhaviv** [comment]
This creates another problem as Lambda is limiting the jar size to 50MB and dragging in the parquet-avro & hadoop-commons dependencies bloat the jar to 64MB (working with the Example connector)

**Comment by danielhaviv** [comment]
I've successfully updated the jar by uploading it to S3 first

**Comment by avirtuos** [comment]
@danielhaviv Can you give more detail about what you did to circumvent the 50MB jar size? I was not aware of any workaround for this other than shrinking the jar size. Additionally, have you checked to see what is actually in the jar? hadoop-commons is probably pulling in considerably more than you need. If you look at the hbase connector in the repo we run some maven dependency exclusion commands to remove significant (>100MB) or unnecessary dependencies that come with the hbase-client (it pulls in large parts of hadoop / hadoop-mr).

**Comment by danielhaviv** [comment]
@avirtuos This is how I'm updating my Lambda function:
```athena-example daniel.haviv$ du -h /Users/daniel.haviv/Documents/projects/aws-athena-query-federation/athena-example/target/athena-example-1.0.jar
 65M	/Users/daniel.haviv/Documents/projects/aws-athena-query-federation/athena-example/target/athena-example-1.0.jar

aws2 s3 cp /Users/daniel.haviv/Documents/projects/aws-athena-query-federation/athena-example/target/athena-example-1.0.jar s3://dhavivresearch/jars/athena-example-1.0.jar
aws2 lambda update-function-code --function-name connector_test --s3-bucket dhavivresearch --s3-key jars/athena-example-1.0.jar
```

You're right regarding the Hadoop deps, I'll have a look at the HBase connector.

Thanks!

**Comment by avirtuos** [comment]
the final command above succeeds? I'd expect it to be subject to the same 50MB limit. Can you confirm the code actually updated? (maybe put a new log line in the code and do the update). I'm just very surprised this works because its basically the same thing our publish tool does and it the failure we saw were within Lambda not in the S3 upload.

**Comment by danielhaviv** [comment]
It succeeds, I verified by adding --publish to the update-function-call and it creates a new RevisionId and Version.

You can also see in this screenshot from the Lamdba console that it is 63 MB:
<img width="1203" alt="Screenshot 2020-03-27 at 18 03 07" src="https://user-images.githubusercontent.com/11798914/77769747-4b88a980-7055-11ea-9eac-019e444708ef.png">

**Comment by atennak1** [comment]
@avirtuos Looks like direct S3 upload is a [known workaround](https://hackernoon.com/exploring-the-aws-lambda-deployment-limits-9a8384b0bec3). But maybe SAM/SAR is enforcing the limit in our script or SAM/SAR uses the Lambda direct upload API.

**Comment by avirtuos** [comment]
The vertica connector (in PR) has a working Parquet reader implementation. It might serve as a good reference / workaround until we support meta-data only connectors.

**Comment by danielhaviv** [comment]
Correct me if I'm wrong, but from what I understand the only implementation
of a Parquet reader is through S3Select

**Comment by avirtuos** [comment]
Nope, you are correct. I'm chatting with the engineer about it right now as it was my understanding we are doing Parquet natively from within the Lambda. I'll update here once we square that away. I wouldn't recommend using S3 Select purely for transcoding as the Vertica connector is presently doing in that PR.

**Comment by danielhaviv** [comment]
@avirtuos Any updates coming soon 🙏?
